Secondary user in-place connection flow

fgiannar opened the pull request:
Fixes #16402 

#### Changes proposed in this Pull Request:
This PR adds in-place connection functionality for secondary users. This means that when a user, that hasn't yet linked their WordPress.com account while the corresponding site is already registered, tries to do so, instead of being redirected to WordPress.com they will be able to link their account in-place, aka without leaving the site at all:
* Adds in-place connection flow in the existing `connect-button` functionality and introduces a new `connectInPlace` property, by default `true`.
* In case the `connectInPlace` property is true, an `iframe` is loaded so that authorization is performed without redirecting the user. An event listener is also attached to `window` in order to detect successful authorization and update the current `state` with the updated linked user data.
* Adds a new client component, `ConnectionBanner` in order to utilize the `ConnectButton` when the user is prompt to link their account via `JetpackNotices`
* Adds a new client component, `AuthIframe`
* Fixes the `connect_url` after a user has unlinked their account
* Ignores the `connectInPlace`, aka triggering the original flow (redirect user) if the detected browser is Safari (both on mobile and desktop) due to the limitation in how 3rd party cookies are handled in Safari
* Replaces the `Admin Notice` message: "Jetpack is powering your site, but to access all of its features you’ll need to create an account." to "Jetpack is powering your site, but to access all of its features you’ll need to connect your account to WordPress.com." and the corresponding CTA text: "Create account" to "Link to WordPress.com"

#### Jetpack product discussion
- p9dueE-1oK-p2

#### Does this pull request change what data or activity we track or use?
The pre-existing functionality of linking a secondary user, included two different options:
1. The user could link their account by clicking the `Create account` button in an admin notice (utilizing `JetpackBanner`)
2. The user could link their account by clicking the `Link to WordPress.com` link in the Jetpack Dashboard, Account connection section.
When 1. was chosen and because of using the `JetpackBanner` the [action was tracked](https://github.com/Automattic/jetpack/blob/master/_inc/client/components/banner/index.jsx#L73).
When 2. was chosen the action was not tracked.

In the current state of this PR the action is in general not tracked.

#### Testing instructions:

#### Prerequisites
- A site already connected to Jetpack.com
- A user account not yet linked to Jetpack.com

- Go to Jetpack-> Dashboard
- Verify the following admin notice is visible:
![Screenshot 2020-06-29 at 4 30 28 PM](https://user-images.githubusercontent.com/1758399/86011882-ecf11400-ba25-11ea-8366-01e51c146db2.png)
- Click the "Link to WordPress.com" button
- Verify the successful WordPress.com in-place authorization
- Unlink your account by clicking the `Unlink me from WordPress.com` link in the Jetpack Dashboard, Account connection section.
- Click the `Link to WordPress.com` link in the same section
- Verify the successful WordPress.com in-place authorization

Make sure to test in different browsers ( Chrome, Firefox, Safari, IE) and devices (desktop, mobile).

- Verify that in Safari desktop or mobile the original flow is triggered and upon linking your account you are redirected to WordPress.com to complete the authorization process there instead.

#### Before
https://d.pr/v/9NlJsd

#### After ( clicking the connect button in the banner )
https://d.pr/v/A2dght

#### After ( clicking the connect button in the Account connection section. )
https://d.pr/v/jwF4m8

#### Proposed changelog entry for your changes:
Secondary user in-place connection flow

fgiannar commented:
Hi @keoshi, many thanks for taking the time to review this PR!

Please find my answers inline:

> * I could see the notice with every type of role, not just Administrator. Not sure if that's supposed to happen because I'm unsure what's “unlocked” when you connect to WP.com, but we should make those match.

I haven't changed anything in the notice display logic, so I guess that would be ok. A confirmation from any crew reviewer would be great though.

> * The “Link to WordPress.com” button breaks line in some screen sizes. What do you think of adding `&nbsp;` to every space so it all stays in a single line?

I am not sure how easily that could be implemented, since we are using the 'i18n-calypso' package to translate all strings and also JSX for rendering elements. I'll look into that though.

> * Seems like the page scrolls to the new element that appears and it gets partially covered by the top admin bar. I wonder if it would be better to just scroll to the very top of the page.

The idea here was to attach the scrolling effect to the iframe itself. This way if we decide eg to move it to the bottom of the page, it should still work.  I have updated it to scroll a bit further up from it's current position, hope this is ok :)

> * The newly added box seems to have an unnecessary amount of bottom padding that could be reduced to visually center the content inside vertically.

Unfortunately, this is not a padding but a set `height` (`220px`) the iframe has. With this height, I've tested it and looks ok both on non-logged in and logged-in (to WP.com) scenarios. If I set it to `100%`, then it displays:
![Screenshot 2020-07-03 at 4 04 47 PM](https://user-images.githubusercontent.com/1758399/86474355-4ff6e980-bd4b-11ea-9d19-8888ab337f86.png)
Any further advice on how to make this work would be more than welcome! 

> * Finally, the transition feels a bit sluggish — I believe it could be a lot faster, 1.5s max.
I've updated it to 1.5s, thanks!

jeherve commented on _inc/client/components/connect-button/index.jsx:
```diff
@@ -8,6 +8,7 @@ import Button from 'components/button';
 import { translate as __ } from 'i18n-calypso';
 import analytics from 'lib/analytics';
 import getRedirectUrl from 'lib/jp-redirect';
+import UAParser from 'ua-parser-js';
```
Instead of introducing a new dependency, I wonder if we could get that data from what we already have in Jetpack. Maybe we could rely on the Device Detection package that’s in Jetpack and pass that data to React here:
https://github.com/Automattic/jetpack/blob/5b9171eeea2ac30d0f5435b7e11d089960ef9999/_inc/lib/admin-pages/class.jetpack-react-page.php#L333
You can then fetch it in `_inc/client/state` and use it

Another alternative would be to rely on the global available in WP:
https://codex.wordpress.org/Global_Variables#Browser_Detection_Booleans

If we do choose to go with the new dependency, you'll want to explicitly declare it in `package.json`; we cannot just assume it will always be available to us as a dependency already there from existing packages.

fgiannar replied:
Many thanks @jeherve for your suggestions.
I have replaced the `ua-parser-js`  based functionality for checking `Safari` browser with an initial state reducer: `isSafari` and take advantage of the global `$is_safari` to pass the data to React.

fgiannar commented:
> Hi @fgiannar.
> I noticed that the flow ignores the constant `JETPACK_SHOULD_NOT_USE_CONNECTION_IFRAME`.
> We could add that into the initial state.

Nice catch, @sergeymitr , thanks! I have updated it accordingly.

fgiannar commented on _inc/client/components/connection-banner/index.jsx:
```diff
@@ -0,0 +1,78 @@
+import ConnectButton from 'components/connect-button';
+import Gridicon from 'components/gridicon';
+
+class ConnectionBanner extends Banner {
```
This corresponding link https://reactjs.org/docs/components-and-props.html#function-and-class-components mentions that ES6 classes are also valid as they are equivalent to function components from React’s point of view.
I used an ES6 class, trying to be consistent with the rest components' definitions (ES6 classes).
I'll make sure to use function components though from now on, thanks :)

jeherve replied:
Classes are definitely valid 👍 That's how we've done things in Jetpack's Dashboard until a few months ago. However, as function components become more and more widely used and preferred, including in projects like Gutenberg, we're trying to head into that direction as well. 

That doesn't mean we need to refactor all our existing work, but that's something to keep in mind when creating new components. 👍

fgiannar commented on _inc/client/state/connection/reducer.js:
```diff
@@ -200,6 +217,16 @@ export function isUnlinkingUser( state ) {
 	return !! state.jetpack.connection.requests.unlinkingUser;
 }
 
+/**
+ * Returns true if currently linking the user
+ *
+ * @param  {Object} state Global state tree
+ * @return {bool} true if currently linking a user, false otherwise
+ */
+export function isAuthorizingUserInPlace( state ) {
```
My approach here was to be explicit about iframe-based authorization, as it is special in the way that we only know that it is happening because the iframe is actually displayed.
Actually, initially I had named it `isLinkingUser` but my issue was that it was not what was actually done behind the scenes, since this is not an explicit API request we can trigger and watch.
The only thing we can be sure about, is that the auth-iframe is displayed that's why I chose to be explicit about it.

fgiannar commented on _inc/client/components/auth-iframe/style.scss:
```diff
@@ -0,0 +1,31 @@
+.fade-in {
+    animation: fadeIn ease 1.5s;
```
Apologies for this, my editor seems to have some issues with `scss` files. I believe it's fixed now.

fgiannar commented:
@jeherve Many thanks for your detailed review and the time you took to review this PR :)

Please find my comments below:
> In some cases the whole page is loaded in the iFrame:

This is strange because while the app is fetching the `connectUrl` the corresponding connect buttons are disabled (click events are not triggered). Furthermore, the iframe source in your screenshot seems to include the `register` path which usually is returned when an account is linked. I tried to reproduce this by quickly linking/unlinking my account but it seemed to work as expected. Any further details on how this was triggered would be much appreciated.

> Could we make the card higher in those scenarios, so the buttons to contact support don't get lost?

Unfortunately, the iframe currently has a standard set `height` as otherwise it displays cropped with a scrollbar. I tried to use a height that displays nice for the most common possible iframe content. I have tried many things as a workaround, including trying to modify its inner height `on load` but this triggers CORS related errors.

> There was no success message upon connection, so I wasn't sure it had worked. Would there be a way to display a success message to the user when the iFrame gets removed?

I have added a success message.

> I wonder if we should try to remove the other elements on the page once you start the connection process. 

In that case, I think it would make more sense to use a new page maybe, dedicated to in place connection. The current approach was to load the iframe on top of every existing page and scroll to it once the user clicks the corresponding button to connect. Not sure about this, I could open a new discussion with the design team though to explore it further :)

> Since we're editing this, I think it would be nice to add tracking to the button click, to keep track of this.

I have added this, thanks.

> Not for this PR, but since we're making changes here; the Connect button at the bottom of the Stats card gets a bit lost in the footer:

I tried placing the connect button next to the `View detailed stats` button, but because it's legend is quite long it was placed under it instead. As a workaround I left it at its current place but made it a button instead of a link so that it catches the eye easier. Hope that's ok :)

jeherve commented on _inc/client/components/auth-iframe/index.jsx:
```diff
@@ -0,0 +1,104 @@
+	};
+
+	static defaultProps = {
+		title: 'Link to WordPress',
```
Should this be translatable?

fgiannar replied:
Thanks, Jeremy! I am currently working on some changes in the iframe title in the [In-place reconnect PR](https://github.com/Automattic/jetpack/pull/16475) so I'll make sure to fix that there.
